HW-DAO6

Max-Pochepets opened the pull request:
DAO6 completed

Max-Pochepets committed:
first

Max-Pochepets committed:
Initial

Mandybur800 commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,169 @@
+            Car car = null;
+            if (set.next()) {
+                car = setCar(set);
+                PreparedStatement statement2 = connection.prepareStatement(getDriversQuery);
```
You can replace it to try-with-resourses

Mandybur800 commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,169 @@
+                    + "m.manufacturer_name, m.manufacturer_country FROM cars c "
+                + "JOIN manufacturers m ON m.manufacturer_id = c.manufacturer_id "
+                + "WHERE c.car_id = ? AND c.deleted = FALSE";
+        String getDriversQuery = "SELECT d.driver_id, d.driver_name, d.driver_license_number "
+                    + "FROM cars c "
+                + "JOIN cars_drivers cd on c.car_id = cd.cars_id "
+                + "JOIN drivers d on d.driver_id = cd.driver_id "
+                + "WHERE c.car_id = ? AND c.deleted = FALSE";
```
You can select FROM cars_drivers and then you can remove JOIN with cars, in WHERE statement you can remove chech for deleted, because it checked in prev. query

Mandybur800 commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,169 @@
+        String updateCar = "UPDATE cars "
+                + "SET car_model = ?, manufacturer_id = ? "
+                + "WHERE car_id = ? AND deleted = FALSE";
+        String deleteRelations = "DELETE cd FROM cars_drivers cd "
+                + "JOIN cars c on c.car_id = cd.cars_id "
+                + "WHERE cars_id = ? AND c.deleted = FALSE";
```
```suggestion
        String deleteRelations = "DELETE FROM cars_drivers "
                + "WHERE cars_id = ?";
```

Mandybur800 commented on src/main/java/crud/dao/impl/DriverDaoJdBcImpl.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public Optional<Driver> get(Long id) {
+        String query = "SELECT driver_id, driver_name, driver_license_number FROM drivers "
+                + "WHERE driver_id = " + id + " AND deleted = FALSE";
+        try (Connection connection = ConnectionUtil.getConnection();
+                Statement statement = connection.createStatement()) {
```
Just use PreparedStatement Because it will be more secure
```suggestion
        String query = "SELECT driver_id, driver_name, driver_license_number FROM drivers "
                + "WHERE driver_id = ? AND deleted = FALSE";
        try (Connection connection = ConnectionUtil.getConnection();
                Statement statement = connection.createStatement()) {
```

Mandybur800 commented on src/main/java/crud/dao/impl/DriverDaoJdBcImpl.java:
```diff
@@ -0,0 +1,113 @@
+        String query = "SELECT * FROM drivers WHERE deleted = FALSE";
+        List<Driver> drivers = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection();
+                Statement statement = connection.createStatement()) {
```
```suggestion
                PreparedStatement statement = connection.createStatement()) {
```

Mandybur800 commented on src/main/java/crud/dao/impl/DriverDaoJdBcImpl.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE drivers SET deleted = TRUE WHERE driver_id = " + id;
```
```suggestion
        String query = "UPDATE drivers SET deleted = TRUE WHERE driver_id = ?";
```

Resci commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,169 @@
+        String query = "SELECT c.car_id FROM cars_drivers cd "
+                + "JOIN cars c on c.car_id = cd.cars_id "
+                + "WHERE cd.driver_id = ? AND c.deleted = FALSE";
+        List<Car> cars = new ArrayList<>();
```
Initialize variables closest to his usage

Resci commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,169 @@
+            ResultSet set = statement.executeQuery();
+            while (set.next()) {
+                Long carId = set.getObject(1, Long.class);
+                cars.add(get(carId).get());
```
If you join cars, why you use get method instead of setCar? (setCar is not the best method name. Call this like "buildCar")

Max-Pochepets replied:
because `setCar()` is already present in `get()`

dmytro-hryhoruk replied:
agree, method get will additionally request which is redundant because you have all information in this resultset

Resci replied:
Використовувати get в getAll погана практика тому, що кожного разу у тебе робиться новий запит в бд і це сильно вдарить по перфомансу

Resci commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,169 @@
+                    + "m.manufacturer_name, m.manufacturer_country FROM cars c "
+                + "JOIN manufacturers m ON m.manufacturer_id = c.manufacturer_id "
+                + "WHERE c.car_id = ? AND c.deleted = FALSE";
+        String getDriversQuery = "SELECT d.driver_id, d.driver_name, d.driver_license_number "
```
This can be done in a separate method

dmytro-hryhoruk replied:
agree let's create method getDrivers() or something like this

Max-Pochepets committed:
Refactor:

in CarDao.get(id) statement2 relocated to try with resources block
in CarDao.get(id) getDriversQuery simplified
in CarDao.update(car) deleteRelations query simplified
in DriverDao Statements changed to PreparedStatements

Resci commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Override
+    public List<Car> getAll() {
```
Same remarks as in getAllByDriver

dmytro-hryhoruk replied:
let's joing the mamufactures here and you will have all information about car. Later you can user getDrivers() method for each of them

dmytro-hryhoruk commented on src/main/java/crud/App.java:
```diff
@@ -10,20 +14,52 @@
     private static final Manufacturer MANUFACTURER_2 = new Manufacturer("Audi", "Germany");
     private static final Manufacturer MANUFACTURER_3 = new Manufacturer("Mercedes", "Germany");
-    private static final String MANUFACTURERS = "manufacturers";
+    private static final Driver DRIVER = new Driver("Ed", "123qwe222113");
+    private static final Driver DRIVER_2 = new Driver("John", "AAAEWWW222JHDSF8899");
+    private static final Driver DRIVER_3 = new Driver("BOB", "IEH709898SDF7689SDF231");
```
don't use numbers in naming. User their real names DRIVER_ED , DRIVER_JOHN and so on

dmytro-hryhoruk commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,166 @@
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, driverId);
+            ResultSet set = statement.executeQuery();
```
```suggestion
            ResultSet resultSet = statement.executeQuery();
```

dmytro-hryhoruk replied:
set it's way too abstract name

dmytro-hryhoruk commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,166 @@
+    public List<Car> getAllByDriver(Long driverId) {
+        String query = "SELECT c.car_id FROM cars_drivers cd "
+                + "JOIN cars c on c.car_id = cd.cars_id "
+                + "WHERE cd.driver_id = ? AND c.deleted = FALSE";
```
what if this driver is  deleted?

dmytro-hryhoruk commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,166 @@
+                + "WHERE cars_id = ?";
+        String updateDrivers = "INSERT INTO cars_drivers (driver_id, cars_id) VALUES (?, ?)";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(updateCar);
+                PreparedStatement statement2 = connection.prepareStatement(deleteRelations);
+                PreparedStatement statement3 = connection.prepareStatement(updateDrivers)) {
```
naming

dmytro-hryhoruk commented on src/main/java/crud/dao/impl/CarDaoJdBcImpl.java:
```diff
@@ -0,0 +1,166 @@
+                statement3.executeUpdate();
+            }
+            return car;
+        } catch (SQLException throwables) {
```
```suggestion
        } catch (SQLException throwable) {
```

dmytro-hryhoruk commented on src/main/java/crud/dao/impl/DriverDaoJdBcImpl.java:
```diff
@@ -0,0 +1,115 @@
+        }
+    }
+
+    static Driver setDriver(ResultSet resultSet) throws SQLException {
```
let's better repeat this code in each class then make this method static and without access modifier

Max-Pochepets committed:
Refactor:

DAO implementations renamed
in CarDaoImpl getAllByDriver() list initialization moved closer to usage
drivers addition moved to separate method
made drivers variables more informative
logic in getAllByDriver improved

Max-Pochepets committed:
Initial

Max-Pochepets committed:
+ empty line

Max-Pochepets committed:
Merge pull request #5

HW-DAO7
